rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,1005 @@
+      } yield r
+
+    for {
+      _              <- initConditionsException.whenA((skipSize, takeSize) == (0, 0))
+      rootNodeSerOpt <- getNodeDataFromStore(rootHash)
+      r              <- rootNodeSerOpt.map(doExport).getOrElse(emptyResult)
+    } yield r
+  }
+
+  /**
+    * Radix Tree implementation
+    */
+  class RadixTreeImpl[F[_]: Sync: Parallel](store: KeyValueTypedStore[F, ByteVector, ByteVector]) {
+
+    /**
+      * Load and decode serializing data from KVDB.
+      */
+    private def loadNodeFromStore(nodePtr: ByteVector): F[Option[Node]] =
+      store.get1(nodePtr).map(_.map(Codecs.decode))
+
+    /**
+      * Cache for storing read and decoded nodes.
+      *
+      * Cache stores kv-pairs (hash, node).
+      * Where hash - Blake2b256Hash of serializing nodes data,
+      *       node - deserialized data of this node.
+      */
+    private val cacheR: TrieMap[ByteVector, Node] = TrieMap.empty
+
+    /**
+      * Load one node from [[cacheR]].
+      *
+      * If there is no such record in cache - load and decode from KVDB, then save to cacheR.
+      * If there is no such record in KVDB - execute assert (if set noAssert flag - return emptyNode).
+      */
+    def loadNode(nodePtr: ByteVector, noAssert: Boolean = false): F[Node] = {
+      def errorMsg(): Unit = assert(noAssert, s"Missing node in database. ptr=${nodePtr.toHex}.")
+      def cacheMiss =
+        for {
+          storeNodeOpt <- loadNodeFromStore(nodePtr)
+          _            = storeNodeOpt.map(cacheR.update(nodePtr, _)).getOrElse(errorMsg())
+        } yield storeNodeOpt.getOrElse(emptyNode)
+      for {
+        cacheNodeOpt <- Sync[F].delay(cacheR.get(nodePtr))
+        r            <- cacheNodeOpt.map(_.pure).getOrElse(cacheMiss)
+      } yield r
+    }
+
+    /**
+      * Clear [[cacheR]] (cache for storing read nodes).
+      */
+    def clearReadCache(): Unit = cacheR.clear()
+
+    /**
+      * Cache for storing serializing nodes. For subsequent unloading in KVDB
+      *
+      * Cache stores kv-pairs (hash, bytes).
+      * Where hash -  Blake2b256Hash of bytes,
+      *       bytes - serializing data of nodes.
+      */
+    private val cacheW: TrieMap[ByteVector, ByteVector] = TrieMap.empty
+
+    /**
+      * Serializing and hashing one [[Node]].
+      *
+      * Serializing data load in [[cacheW]].
+      * If detected collision with older cache data - executing assert
+      */
+    def saveNode(node: Node): ByteVector = {
+      val (hash, bytes) = hashNode(node)
+      def checkCollision(v: Node): Unit = assert(
+        v == node,
+        s"Collision in cache: record with key = ${hash.toHex} has already existed."
+      )
+      cacheR.get(hash).map(checkCollision).getOrElse(cacheR.update(hash, node))
+      cacheW.update(hash, bytes)
+      hash
+    }
+
+    /**
+      * Save all cacheW to KVDB
+      *
+      * If detected collision with older KVDB data - execute Exception
+      */
+    def commit(): F[Unit] = {
```

Comment:
This function can be defined without parenthesis. When called `F[Unit]` will be constructed as an expression and not executed immediately.